contracts/TokenListManager.sol

```diff
@@ -2,96 +2,217 @@
 
     // Use mapping to check if token in whitelist
     mapping(address => bool) private whitelistedTokens;
     // Use mapping to check if token in blacklist
     mapping(address => bool) private blacklistedTokens;
 
+    // Set permissions
+    modifier onlyOwner() {
+        // Match address
+        /** 0xC0Ec75c25EC201885a791Fd8d39Bf8CE96e1c566
+         * this address is my own test address
+         * owner can use thier own address to replace. */
+        require(msg.sender == 0xC0Ec75c25EC201885a791Fd8d39Bf8CE96e1c566, "Not authorized");
+        _;
+    }
+
     // Implement the addTokens method and add tokens to whitelist
-    function addTokens(address[] calldata tokens) external override {
+    function addTokens(address[] calldata tokens) external override onlyOwner {
         for (uint i = 0; i < tokens.length; i++) {
-            // Mark each token address as whitelisted
-            whitelistedTokens[tokens[i]] = true;
+            // Check if the token is not already whitelisted
+            if (!whitelistedTokens[tokens[i]]) {
+                // If the token type is ERC-20
+                if (IERC165(tokens[i]).supportsInterface(type(IERC20).interfaceId)) {
+                    // Storage to whitelistErc20Tokens
+                    whitelistErc20Tokens.push(tokens[i]);
+                // If the token type is IERC-721
+                } else if (IERC165(tokens[i]).supportsInterface(type(IERC721).interfaceId)) {
+                    // Storage to whitelistErc721Tokens
+                    whitelistErc721Tokens.push(tokens[i]);
+                // If the token type is IERC-1155
+                } else if (IERC165(tokens[i]).supportsInterface(type(IERC1155).interfaceId)) {
+                    // Storage to whitelistErc1155Tokens
+                    whitelistErc1155Tokens.push(tokens[i]);
+                } else {
+                    revert("Unsupported token type");
+                }
+                // Mark the token address as whitelisted in the mapping
+                whitelistedTokens[tokens[i]] = true;
+                // Add the token address to the array of whitelisted tokens
+                whitelistedTokensArray.push(tokens[i]);
+            }
         }
     }
 
     // Implement the removeTokens method and remove tokens from whitelist
-    function removeTokens(address[] calldata tokens) external override {
+    function removeTokens(address[] calldata tokens) external override onlyOwner {
         for (uint i = 0; i < tokens.length; i++) {
-            // Remove each token address from whitelist
-            whitelistedTokens[tokens[i]] = false;
+            // Check if the token in whitelist
+            if (whitelistedTokens[tokens[i]]) {
+                // Remove each token address from whitelist
+                whitelistedTokens[tokens[i]] = false;
+                // Call internal function below to remove from the array
+                removeTokenFromWhitelistArray(tokens[i]);
```

Comment:
not removed in erc20/721/1155 whitelist array.

Reply:
Solved